### PR TITLE
GH-118095: Add dynamic exit support and FOR_ITER_GEN support to tier 2

### PR DESCRIPTION
**Pull request by markshannon**
* Issue: gh-118095

### CODE REVIEW

File: Python/bytecodes.c [hunk 1]
```diff
@@ -2759,24 +2763,25 @@ dummy_func(
...
             _PyFrame_StackPush(gen_frame, Py_None);
             gen->gi_frame_state = FRAME_EXECUTING;
             gen->gi_exc_state.previous_item = tstate->exc_info;
             tstate->exc_info = &gen->gi_exc_state;
-            assert(next_instr[oparg].op.code == END_FOR ||
-                   next_instr[oparg].op.code == INSTRUMENTED_END_FOR);
-            assert(next_instr - this_instr + oparg <= UINT16_MAX);
-            frame->return_offset = (uint16_t)(next_instr - this_instr + oparg);
-            DISPATCH_INLINED(gen_frame);
+            frame->return_offset = (uint16_t)(2 + oparg);
```
Comment (gvanrossum) [inline Python/bytecodes.c]: What does `2` stand for? Can you add a comment?

File: Python/bytecodes.c [hunk 2]
```diff
@@ -3169,7 +3174,7 @@ dummy_func(
         // The 'unused' output effect represents the return value
         // (which will be pushed when the frame returns).
         // It is needed so CALL_PY_EXACT_ARGS matches its family.
```
Comment (gvanrossum) [inline Python/bytecodes.c]: This comment is outdated -- I suggest just deleting it.

File: Tools/jit/template.c [hunk 3]
```diff
@@ -87,6 +87,7 @@ _JIT_ENTRY(_PyInterpreterFrame *frame, PyObject **stack_pointer, PyThreadState *
     PATCH_VALUE(_PyExecutorObject *, current_executor, _JIT_EXECUTOR)
     int oparg;
     int uopcode = _JIT_OPCODE;
+    _Py_CODEUNIT *next_instr;
```
Comment (gvanrossum) [inline Tools/jit/template.c:90]: I'd like to know what @brandtbucher thinks of this.

File: Python/bytecodes.c [hunk 4]
```diff
@@ -4189,6 +4194,37 @@ dummy_func(
...
+            _Py_CODEUNIT *target = frame->instr_ptr;
+            _PyExecutorObject *executor;
+            if (target->op.code == ENTER_EXECUTOR) {
+                PyCodeObject *code = (PyCodeObject *)frame->f_executable;
+                executor = code->co_executors->executors[target->op.arg];
+                Py_INCREF(executor);
+            }
+            else {
+                if (!backoff_counter_triggers(exit->temperature)) {
+                    GOTO_TIER_ONE(target);
```
Comment (gvanrossum) [inline Python/bytecodes.c]: This doesn't advance the temperature. From `_SIDE_EXIT`, it looks like the correct idiom is
```
            _Py_BackoffCounter temperature = exit->temperature;
            if (!backoff_counter_triggers(temperature)) {
                exit->temperature = advance_backoff_counter(temperature);
                GOTO_TIER_ONE(target);
            }
```

File: Python/bytecodes.c [hunk 5]
```diff
@@ -4222,6 +4258,7 @@ dummy_func(
             GOTO_UNWIND();
         }
 
+
```
Comment (gvanrossum) [inline Python/bytecodes.c:4260]: Stray blank line?

**Review by gvanrossum — APPROVED** [review]

**Comment by markshannon** [comment]
The usual four failures.

**Comment by gvanrossum** [comment]
@markshannon I now get this warning:
```
In file included from Python/optimizer_analysis.c:439:
Python/optimizer_cases.c.h:1444:23: warning: incompatible pointer types assigning to '_PyInterpreterFrame *' (aka 'struct _PyInterpreterFrame *') from '_Py_UopsSymbol *' (aka 'struct _Py_UopsSymbol *') [-Wincompatible-pointer-types]
            gen_frame = sym_new_not_null(ctx);
                      ^ ~~~~~~~~~~~~~~~~~~~~~
1 warning generated.
```
That's in here:
```c
        case _FOR_ITER_GEN_FRAME: {
            _PyInterpreterFrame *gen_frame;
            gen_frame = sym_new_not_null(ctx);
            if (gen_frame == NULL) goto out_of_space;
            stack_pointer[0] = (_Py_UopsSymbol *)gen_frame;
            stack_pointer += 1;
            break;
        }
```
Apparently generated by the generator using the default code generation strategy. We may need to add a dummy case for `_FOR_ITER_GEN_FRAME` in optimizer_bytecodes.c to avoid this cast.

(@Fidget-Spinner Or is there another way to fix this?)

**Comment by Fidget-Spinner** [comment]
The optimizer bytecodes shouldnt even use the cast. I should fix that because ifs interfering with tagged ptrs too

**Comment by Fidget-Spinner** [comment]
For now, we can just override it I think as a hotfix.

**Pull request merged**